base/linalg/diagonal.jl

```diff
@@ -104,31 +104,85 @@ end
-            return $uplo(B)
-        end
-    end
-end
-(*)(A::AbstractTriangular, D::Diagonal) = error("this method should never be reached")
-(*)(D::Diagonal, A::AbstractTriangular) = error("this method should never be reached")
+
+(*)(A::AbstractTriangular, D::Diagonal) = A_mul_B!(copy(A), D)
+(*)(D::Diagonal, B::AbstractTriangular) = A_mul_B!(D, copy(B))
 
 (*)(A::AbstractMatrix, D::Diagonal) =
     scale!(similar(A, promote_op(*, eltype(A), eltype(D.diag))), A, D.diag)
 (*)(D::Diagonal, A::AbstractMatrix) =
     scale!(similar(A, promote_op(*, eltype(A), eltype(D.diag))), D.diag, A)
 
+A_mul_B!(A::Union{LowerTriangular,UpperTriangular}, D::Diagonal) =
+    typeof(A)(A_mul_B!(A.data, D))
+function A_mul_B!(A::UnitLowerTriangular, D::Diagonal)
+    A_mul_B!(A.data, D)
+    for i = 1:size(A, 1)
+        A.data[i,i] = D.diag[i]
+    end
+    LowerTriangular(A.data)
+end
+function A_mul_B!(A::UnitUpperTriangular, D::Diagonal)
+    A_mul_B!(A.data, D)
+    for i = 1:size(A, 1)
+        A.data[i,i] = D.diag[i]
+    end
+    UpperTriangular(A.data)
+end
+function A_mul_B!(D::Diagonal, B::UnitLowerTriangular)
+    A_mul_B!(D, B.data)
+    for i = 1:size(B, 1)
+        B.data[i,i] = D.diag[i]
+    end
+    LowerTriangular(B.data)
+end
+function A_mul_B!(D::Diagonal, B::UnitUpperTriangular)
+    A_mul_B!(D, B.data)
+    for i = 1:size(B, 1)
+        B.data[i,i] = D.diag[i]
+    end
+    UpperTriangular(B.data)
+end
+
+Ac_mul_B(A::AbstractTriangular, D::Diagonal) = A_mul_B!(ctranspose(A), D)
+function Ac_mul_B(A::AbstractMatrix, D::Diagonal)
+    Ac = similar(A, promote_op(*, eltype(A), eltype(D.diag)), (size(A, 2), size(A, 1)))
+    ctranspose!(Ac, A)
+    A_mul_B!(Ac, D)
+end
+
+At_mul_B(A::AbstractTriangular, D::Diagonal) = A_mul_B!(transpose(A), D)
+function At_mul_B(A::AbstractMatrix, D::Diagonal)
+    Ac = similar(A, promote_op(*, eltype(A), eltype(D.diag)), (size(A, 2), size(A, 1)))
+    transpose!(Ac, A)
+    A_mul_B!(Ac, D)
+end
+
+A_mul_Bc(D::Diagonal, B::AbstractTriangular) = A_mul_B!(D, ctranspose(B))
+A_mul_Bc(D::Diagonal, Q::Union{Base.LinAlg.QRCompactWYQ,Base.LinAlg.QRPackedQ}) = A_mul_Bc!(Array(D), Q)
+function A_mul_Bc(D::Diagonal, A::AbstractMatrix)
+    Ac = similar(A, promote_op(*, eltype(A), eltype(D.diag)), (size(A, 2), size(A, 1)))
+    ctranspose!(Ac, A)
+    A_mul_B!(D, Ac)
+end
+
+A_mul_Bt(D::Diagonal, B::AbstractTriangular) = A_mul_B!(D, transpose(B))
+function A_mul_Bt(D::Diagonal, A::AbstractMatrix)
+    Ac = similar(A, promote_op(*, eltype(A), eltype(D.diag)), (size(A, 2), size(A, 1)))
+    ctranspose!(Ac, A)
+    A_mul_B!(D, Ac)
+end
+
+A_mul_B!(A::Diagonal,B::Diagonal) = throw(ArgumentError("mutation argument is ambiguous. Please promote one of the arguments."))
+At_mul_B!(A::Diagonal,B::Diagonal)= throw(ArgumentError("mutation argument is ambiguous. Please promote one of the arguments."))
+Ac_mul_B!(A::Diagonal,B::Diagonal)= throw(ArgumentError("mutation argument is ambiguous. Please promote one of the arguments."))
```

Comment:
Improving consistency here is beyond the scope of this PR, sure. But this operation isn't an error.